### ISSUE TITLE
[BUG] Can't set theme or background color [Ubuntu 22.04]

### ISSUE THREAD

**Issue by beville** [realmazharhussain/gdm-settings#40]
### What happened?

Trying to set the theme and background color failed.

### Expected behavior

Goal is try to change the background to solid color, and change the theme so that the text entry is white background with blue highlight.

### To Reproduce

Ran gdm-settings, 

Under **Appearance**, I set the shell to Yaru-blue, the icons to Yaru.  Set the background to a light blue color.

Under **Miscellaneous**, I added a welcome message, and disabled the logo.  

Clicked Apply, and logged out.

The login screen didn't change color, and was still using a dark theme.  However, the logo was gone, and the the custom message showed, so I know the app did *something*

### Application Version

0.5.2

### Source

did a git clone

### Environment

Ubuntu 22.04

**Comment by squalou** [comment]
If I may :I face the same issue.
Running flatpak from terminal I get this error when applying changes

```
Traceback (most recent call last):
  File "/app/lib/python3.9/site-packages/gdm_settings/gdm_settings_gui.py", line 95, in on_apply
    if self.settings.apply_settings():
  File "/app/lib/python3.9/site-packages/gdm_settings/settings_manager.py", line 658, in apply_settings
    self.apply_gresource_settings()
  File "/app/lib/python3.9/site-packages/gdm_settings/settings_manager.py", line 556, in apply_gresource_settings
    gresource_utils.auto_backup()
  File "/app/lib/python3.9/site-packages/gdm_settings/settings_manager.py", line 254, in auto_backup
    self._extract_default_pure_theme()
  File "/app/lib/python3.9/site-packages/gdm_settings/settings_manager.py", line 241, in _extract_default_pure_theme
    self.extract_default_theme(target_dir=TEMP_DIR, name='default-pure')
  File "/app/lib/python3.9/site-packages/gdm_settings/settings_manager.py", line 230, in extract_default_theme
    copytree(source_shell_dir, target_shell_dir)
  File "/usr/lib/python3.9/shutil.py", line 564, in copytree
    with os.scandir(src) as itr:
FileNotFoundError: [Errno 2] No such file or directory: '/home/<my_username>/.var/app/io.github.realmazharhussain.GdmSettings/cache/gdm-settings/extracted'

```

**Comment by realmazharhussain** [comment]
@beville Just fixed this issue. Do a `git pull`, and re-compile/install the application.

Don't forget to report back.

**Comment by realmazharhussain** [comment]
@squalou 

> If I may :I face the same issue. Running flatpak from terminal I get this error when applying changes
> 
> ```
> Traceback (most recent call last):
>   File "/app/lib/python3.9/site-packages/gdm_settings/gdm_settings_gui.py", line 95, in on_apply
>     if self.settings.apply_settings():
>   File "/app/lib/python3.9/site-packages/gdm_settings/settings_manager.py", line 658, in apply_settings
>     self.apply_gresource_settings()
>   File "/app/lib/python3.9/site-packages/gdm_settings/settings_manager.py", line 556, in apply_gresource_settings
>     gresource_utils.auto_backup()
>   File "/app/lib/python3.9/site-packages/gdm_settings/settings_manager.py", line 254, in auto_backup
>     self._extract_default_pure_theme()
>   File "/app/lib/python3.9/site-packages/gdm_settings/settings_manager.py", line 241, in _extract_default_pure_theme
>     self.extract_default_theme(target_dir=TEMP_DIR, name='default-pure')
>   File "/app/lib/python3.9/site-packages/gdm_settings/settings_manager.py", line 230, in extract_default_theme
>     copytree(source_shell_dir, target_shell_dir)
>   File "/usr/lib/python3.9/shutil.py", line 564, in copytree
>     with os.scandir(src) as itr:
> FileNotFoundError: [Errno 2] No such file or directory: '/home/<my_username>/.var/app/io.github.realmazharhussain.GdmSettings/cache/gdm-settings/extracted'
> ```

This is quite a different issue. Please! open a new issue with information like Application Version, OS name and version, GNOME version, Where did you get the application, etc.

**Comment by squalou** [comment]
ok, sorry for the hijack :-) opening a new one asap

**Comment by realmazharhussain** [comment]
Fixed the issue. update the application and it should work fine now.

**Comment by rob83gh** [comment]
I just tried version 0.5.3 from flathub. Saving changes (#43) now works, but it still doesn't let me set a background color or picture. No error, it just doesn't change. With the 0.5.3 AppImage it does work! Ubuntu 22.04.

**Comment by realmazharhussain** [comment]
Are you saying that you cannot change background using the Flatpak but you can change it using the AppImage?

**Comment by beville** [comment]
I haven’t had the time to go back and test this again, but a few days after I created this issue, I used your command line tool to set the color.  After I did that, I was able to use GUI to change the color again.  So, maybe the command-line tool was doing some extra setup work?

**Comment by realmazharhussain** [comment]
@rob83gh, is it working for you too?

**Comment by rob83gh** [comment]
> Are you saying that you cannot change background using the Flatpak but you can change it using the AppImage?

Yes, that is what I am saying. Here's what I did, exactly.

1. fresh install of ubuntu 22.04
2. enabled flatpak and installed version 0.5.3
3. used it to change logo and background
4. rebooted

The logo has changed, the background has not.

5. downloaded the AppImage
6. installed the runtime dependencies listed in the readme
7. used that to change the logo and background
8. rebooted

Both changes were applied.

So, I thought maybe the appimage or its dependencies did something extra but no - went back to the flatpak and tried again, it still doesn't work. I was unaware of a command line option, how would I try that?

**Comment by realmazharhussain** [comment]
I'll look into that.

> I was unaware of a command line option, how would I try that?

Before I wrote this application, I wrote 2 bash scripts which did the same thing. I called them [`gdm-tools`](https://github.com/realmazharhussain/gdm-tools). They are not part of this application. They have their own separate GitHub repository at <https://github.com/realmazharhussain/gdm-tools>.

**Comment by rob83gh** [comment]
Can confirm, after using the cli tool to change background, the gui tool also started working.

**Comment by realmazharhussain** [comment]
@rob83gh 

> Can confirm, after using the cli tool to change background, the gui tool also started working.

Even the flatpak?

**Comment by realmazharhussain** [comment]
I don't know what to do here. The problem mentioned in [this comment](#issuecomment-1120498535) seems so random and rare. It shouldn't even exist. Everything GDM background related happens inside `gnome-shell-theme.gresource`, both of these projects use only the default `gnome-shell-theme.gresource` file and their process does not interact with other files on the system (except, of course, for the image chosen as GDM background and the selected shell theme). So, there should be no way for gdm-tools to affect the functionality of this app. But it did affect it.

So, any suggestions?

I was thinking, since the problem mentioned in [this comment](#issuecomment-1120498535) is somewhat different from this issue, I should close this issue and if anyone has this problem again they could create a fresh issue. Any comment on this?

**Comment by rob83gh** [comment]
I think I've figured it out. The bash script runs update-alternatives for gdm-theme.gresource but the flatpak does not so it's still pointing to Yaru theme:

```
There are 2 choices for the alternative gdm-theme.gresource (providing /usr/share/gnome-shell/gdm-theme.gresource).

  Selection    Path                                                           Priority   Status
------------------------------------------------------------
* 0            /usr/share/gnome-shell/theme/Yaru/gnome-shell-theme.gresource   15        auto mode
  1            /usr/share/gnome-shell/gnome-shell-theme.gresource              10        manual mode
  2            /usr/share/gnome-shell/theme/Yaru/gnome-shell-theme.gresource   15        manual mode
```

After running update-alternatives and selecting option 1 myself, I now see the changes I made through the flatpak.

**Comment by realmazharhussain** [comment]
The flatpak is also supposed to do that. I don't know why it's not.

What are the permissions of `/usr/share/gnome-shell/theme/Yaru/gnome-shell-theme.gresource`?

**Comment by rob83gh** [comment]
```
-rw-r--r-- 1 root root 3104735 Apr 13 16:55 /usr/share/gnome-shell/theme/Yaru/gnome-shell-theme.gresource
```

**Comment by realmazharhussain** [comment]
What does running the following commands output?

```bash
flatpak run --command=sh io.github.realmazharhussain.GdmSettings
cd /var/run/host/usr/share/gnome-shell
ls -l *.gresource | sed -E 's/ +/ /g' | cut -d' ' -f 1,9
```

**Comment by rob83gh** [comment]
```
lrwxrwxrwx gdm-theme.gresource
-rw-r--r-- gnome-shell-dbus-interfaces.gresource
-rw-r--r-- gnome-shell-icons.gresource
-rw-r--r-- gnome-shell-osk-layouts.gresource
-rw-r--r-- gnome-shell-theme.gresource
-rw-r--r-- org.gnome.Extensions.data.gresource
-rw-r--r-- org.gnome.Extensions.src.gresource
-rw-r--r-- org.gnome.ScreenSaver.src.gresource
-rw-r--r-- org.gnome.Shell.Extensions.src.gresource
-rw-r--r-- org.gnome.Shell.Notifications.src.gresource
-rw-r--r-- org.gnome.Shell.Screencast.src.gresource
```

**Comment by realmazharhussain** [comment]
Yes, it is helpful.

I will, now, add logging information in the app. It will take time but once it is done, I will upload the flatpak again. This way, it would be much easier to know what is going on in the app.

**Comment by realmazharhussain** [comment]
I have added logging support in the app. Here's the new flatpak

[gdm-settings.flatpak.zip](https://github.com/realmazharhussain/gdm-settings/files/8867451/gdm-settings.flatpak.zip)

After installing it,

- Run the following command in order to enable logging/verbose-mode.

  ```bash
  flatpak run io.github.realmazharhussain.GdmSettings --verbose
  ```
- Apply your settings
- Provide its output here.

**Comment by rob83gh** [comment]
```
flatpak run io.github.realmazharhussain.GdmSettings --verbose
INFO: PackageType = Flatpak
INFO: TEMP_DIR = '/home/rob/.var/app/io.github.realmazharhussain.GdmSettings/cache/tmp'
INFO: self.ShellGresourceFile = '/usr/share/gnome-shell/gnome-shell-theme.gresource'
INFO: self.UbuntuGdmGresourceFile = None
INFO: Applying GResource settings for non-Ubuntu systems ...
```

**Comment by realmazharhussain** [comment]
Does this flatpak have any different output?

[gdm-settings.flatpak.zip](https://github.com/realmazharhussain/gdm-settings/files/8868808/gdm-settings.flatpak.zip)

**Comment by rob83gh** [comment]
No, it is exactly the same. I double checked and made sure I only had the newest flatpak installed.

**Comment by realmazharhussain** [comment]
Is your `org.gnome.Platform` flatpak runtime up-to-date?

**Comment by realmazharhussain** [comment]
What output does [this flatpak](https://github.com/realmazharhussain/gdm-settings/files/8869780/gdm-settings.flatpak.zip) give?

**Comment by rob83gh** [comment]
> Is your `org.gnome.Platform` flatpak runtime up-to-date?

Yes, I believe so:

```
flatpak update
Looking for updates...
Nothing to do.
```
> What output does this flatpak give?

```
flatpak run io.github.realmazharhussain.GdmSettings --verbose
INFO: PackageType = Flatpak
INFO: TEMP_DIR = '/home/rob/.var/app/io.github.realmazharhussain.GdmSettings/cache/tmp'
DEBUG: Files in Shell Dir: ['theme', 'org.gnome.Shell.Extensions', 'gir-1.0', 'org.gnome.ScreenSaver', 'perf-background.xml', 'extensions', 'org.gnome.Extensions.data.gresource', 'search-providers', 'gnome-shell-dbus-interfaces.gresource', 'gdm-theme.gresource', 'Shell-0.1.gir', 'gnome-shell-osk-layouts.gresource', 'org.gnome.Shell.Screencast.src.gresource', 'org.gnome.ScreenSaver.src.gresource', 'gnome-shell-theme.gresource', 'org.gnome.Shell.Extensions.src.gresource', 'org.gnome.Shell.Notifications', 'modes', 'org.gnome.Shell.Screencast', 'org.gnome.Shell.Notifications.src.gresource', 'St-1.0.gir', 'org.gnome.Extensions', 'gnome-shell-icons.gresource', 'gnome-shell-theme.gresource.default', 'org.gnome.Extensions.src.gresource', 'Gvc-1.0.gir']
INFO: self.ShellGresourceFile = '/usr/share/gnome-shell/gnome-shell-theme.gresource'
INFO: self.UbuntuGdmGresourceFile = None
INFO: Applying GResource settings for non-Ubuntu systems ...
```

**Comment by realmazharhussain** [comment]
Try this flatpak.

[gdm-settings.flatpak.zip](https://github.com/realmazharhussain/gdm-settings/files/8869874/gdm-settings.flatpak.zip)

**Comment by rob83gh** [comment]
```
INFO: PackageType = Flatpak
INFO: TEMP_DIR = '/home/rob/.var/app/io.github.realmazharhussain.GdmSettings/cache/tmp'
INFO: self.ShellGresourceFile = '/usr/share/gnome-shell/gnome-shell-theme.gresource'
INFO: self.UbuntuGdmGresourceFile = '/usr/share/gnome-shell/gdm-theme.gresource'
INFO: Applying GResource settings for Ubuntu ...
```

That works! After logging out, I see my changes now. Thanks! For your time and for the program.

**Comment by realmazharhussain** [comment]
Nah, man. Thank YOU. Usually, people don't keep helping the developers diagnose/fix the issue for that long. But you did. So, thank you.
